packages/flutter/lib/src/material/dropdown.dart

```diff
@@ -681,21 +684,40 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
   /// Creates an item for a dropdown menu.
   ///
   /// The [child] argument is required.
+  ///
+  /// If [enabled] is `false`, [value] and [onTap] should be null.
   const DropdownMenuItem({
     Key? key,
     this.onTap,
     this.value,
+    this.enabled = true,
     required Widget child,
   }) : assert(child != null),
+       assert(enabled || (!enabled && value == null), 'if enabled is false, value should be null'),
+       assert(enabled || (!enabled && onTap == null), 'if enabled is false, onTap should be null'),
        super(key: key, child: child);
 
   /// Called when the dropdown menu item is tapped.
+  ///
+  /// If [enabled] is `false`, this property should be `null`.
   final VoidCallback? onTap;
 
   /// The value to return if the user selects this menu item.
   ///
   /// Eventually returned in a call to [DropdownButton.onChanged].
+  ///
+  /// If [enabled] is `false`, this property should be `null`.
```

Comment:
Why is this a requirement? As long as it is disabled, the value is irrelevant.

Reply:
This was suggested by @shihaohong sir in [this](https://github.com/flutter/flutter/pull/76968#discussion_r589275180) conversation. If you feel otherwise, I'll change it.

Reply:
The rationale I had here was that it might be confusing to developers. For example, if someone had (unknowingly) disabled a menu item and then defined the `value` or `onTap`, then the irrelevant values are still defined in their code but have no effect on the widget itself, since it's disabled. I was thinking it might be better to catch this in an assert ahead of time than allow devs to scratch their heads trying to figure it out.

Reply:
Well, but it also means that they have to have extra logic. Instead of just setting a static value on the widget, they now need to have `enabled ? value : null` instead. Same with the onTap, although there it makes a little more sense to assert, since they might be confused as to why their onTap is never called when its disabled. It's still extra logic that they have to provide now, since it'll assert otherwise. They might not care that there's an onTap defined that is only called when enabled is true.

So, instead of what would be consistent with other widgets:
```dart
DropdownMenuItem(
  // ...
  onTap: myEnabled ? () { /*...*/ } : null,
  value: foo,
)
```

They have to write the (fairly redundant):
```dart
DropdownMenuItem(
  // ...
  onTap: myEnabled ? () { /*...*/ } : null,
  value: myEnabled ? foo : null,
  enabled: myEnabled,
)
```

And because of the asserts, they don't have the option of the more readable:
```
DropdownMenuItem(
  // ...
  onTap: () { /*...*/ },
  value: foo,
  enabled: myEnabled,
)
```

However, I realize that since we didn't have the consistent API from the beginning, it's pretty common to have a dropdown item with no `onTap` set now, so I guess we can't get away with the breaking change, but maybe we can skip the asserts?

Reply:
Instead of assigning the disability of the item with `onTap` being `null`, shouldn't it be done with `value` being `null`? i.e., if `value` is `null`, item will be disabled, else enabled. Right now, item's value being `null` has no significance. This would give it that, and would keep the pattern as with other widgets.

For `onTap`, since it is not directly related to the basic working of `DropdownButton`, we can excempt it from being dependent on item being enabled or not, i.e., `onTap` will work irrespective of item being disabled or enabled (or `value` being `null` or not).

Reply:
> Instead of assigning the disability of the item with onTap being null, shouldn't it be done with value being null?

@werainkhatri That would actually be a breaking change because I recently worked through some issues with users expecting `null` values to be valid (so to these subset of users, `null` value do have significance).

Let's remove the asserts. I think @gspencergoog's concerns about users requiring additional logic in their code just to disable their dropdown menu items makes a lot of sense

Reply:
Removed the asserts.

I also came across a bug which I fixed.
I went through some other related PRs and realized that if `DropdownButton.value` is `null` and there is an item with `value` as null, that item is selected. But if there are >= 1 disabled items along with that, it causes an assertion error as they are considered similar. I changed the condition to selecting the item if its `value` is `null` and its not disabled.